File: analyze_constants.py
#!/usr/bin/env python3
"""
Analyze the Windows API constants file to understand categorization patterns.
"""

import re
from collections import defaultdict, Counter

def analyze_constants_file(filepath):
    """Analyze the constants file and categorize by prefixes."""
    
    constants = []
    prefixes = defaultdict(list)
    
    # Read the file and extract constants
    with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
        content = f.read()
    
    # Find all add_const calls
    pattern = r"win_const_mgr\.add_const\('([^']+)',\s*0x[0-9A-Fa-f]+\)"
    matches = re.findall(pattern, content)
    
    print(f"Found {len(matches)} constants")
    
    # Categorize by prefixes
    for const_name in matches:
        constants.append(const_name)
        
        # Extract prefix (everything before first underscore or first digit)
        if '_' in const_name:
            prefix = const_name.split('_')[0]
        else:
            # For constants without underscore, take first few chars
            prefix = const_name[:3] if len(const_name) > 3 else const_name
            
        prefixes[prefix].append(const_name)
    
    # Count prefixes
    prefix_counts = Counter({k: len(v) for k, v in prefixes.items()})
    
    print("\nTop 50 prefixes by count:")
    for prefix, count in prefix_counts.most_common(50):
        print(f"{prefix:20} : {count:5} constants")
        
    print(f"\nTotal prefixes: {len(prefix_counts)}")
    print(f"Total constants: {len(constants)}")
    
    # Look for logical groupings
    print("\n=== Suggested Logical Groupings ===")
    
    # Error codes
    error_prefixes = [p for p in prefix_counts if 'ERROR' in p or 'ERR' in p]
    error_count = sum(prefix_counts[p] for p in error_prefixes)
    print(f"Error codes ({len(error_prefixes)} prefixes): {error_count} constants")
    
    # Window messages
    wm_prefixes = [p for p in prefix_counts if p.startswith('WM') or 'MSG' in p]
    wm_count = sum(prefix_counts[p] for p in wm_prefixes)
    print(f"Window messages ({len(wm_prefixes)} prefixes): {wm_count} constants")
    
    # Registry
    reg_prefixes = [p for p in prefix_counts if 'HKEY' in p or 'REG' in p or 'KEY' in p]
    reg_count = sum(prefix_counts[p] for p in reg_prefixes)
    print(f"Registry ({len(reg_prefixes)} prefixes): {reg_count} constants")
    
    # Security/Certificates
    sec_prefixes = [p for p in prefix_counts if any(x in p for x in ['CERT', 'CRYPT', 'SEC', 'AUTH', 'TRUST'])]
    sec_count = sum(prefix_counts[p] for p in sec_prefixes)
    print(f"Security/Crypto ({len(sec_prefixes)} prefixes): {sec_count} constants")
    
    # File system
    file_prefixes = [p for p in prefix_counts if any(x in p for x in ['FILE', 'DIR', 'DRIVE', 'VOLUME', 'DISK'])]
    file_count = sum(prefix_counts[p] for p in file_prefixes)
    print(f"File system ({len(file_prefixes)} prefixes): {file_count} constants")
    
    # Network/HTTP
    net_prefixes = [p for p in prefix_counts if any(x in p for x in ['HTTP', 'NET', 'DNS', 'TCP', 'UDP', 'IP', 'SOCKET', 'WSAE'])]
    net_count = sum(prefix_counts[p] for p in net_prefixes)
    print(f"Network ({len(net_prefixes)} prefixes): {net_count} constants")
    
    # Database/SQL
    db_prefixes = [p for p in prefix_counts if 'SQL' in p or 'DB' in p]
    db_count = sum(prefix_counts[p] for p in db_prefixes)
    print(f"Database ({len(db_prefixes)} prefixes): {db_count} constants")
    
    # Graphics/DirectX
    gfx_prefixes = [p for p in prefix_counts if any(x in p for x in ['DD', 'D3D', 'GDI', 'DIB', 'BMP', 'IMAGE'])]
    gfx_count = sum(prefix_counts[p] for p in gfx_prefixes)
    print(f"Graphics ({len(gfx_prefixes)} prefixes): {gfx_count} constants")
    
    return prefixes, prefix_counts

if __name__ == "__main__":
    filepath = "/workspace/lib/rex/post/meterpreter/extensions/stdapi/railgun/def/windows/api_constants.rb"
    analyze_constants_file(filepath)

File: analyze_constants.rb
#!/usr/bin/env ruby

# Script to analyze Windows API constants and categorize them for splitting

require 'set'

# Read the constants file
constants_file = '/workspace/lib/rex/post/meterpreter/extensions/stdapi/railgun/def/windows/api_constants.rb'
content = File.read(constants_file)

# Extract all constant names
constants = []
content.scan(/win_const_mgr\.add_const\('([^']+)'/) do |match|
  constants << match[0]
end

puts "Total constants found: #{constants.length}"

# Categorize constants by prefix patterns
categories = Hash.new { |h, k| h[k] = [] }

constants.each do |const|
  case const
  when /^ERROR_/
    categories['errors'] << const
  when /^WM_/, /^HWND_/, /^WS_/, /^SW_/, /^SWP_/, /^VK_/, /^IDC_/, /^IDI_/, /^MB_/, /^IDOK/, /^IDCANCEL/, /^IDYES/, /^IDNO/
    categories['ui_windows'] << const
  when /^REG/, /^HKEY_/, /^KEY_/
    categories['registry'] << const
  when /^FILE_/, /^GENERIC_/, /^CREATE_/, /^OPEN_/, /^SHARE_/
    categories['file_io'] << const
  when /^PROCESS_/, /^THREAD_/, /^TOKEN_/, /^SE_/
    categories['process_security'] << const
  when /^SERVICE_/, /^SC_/
    categories['services'] << const
  when /^DNS_/, /^SOCKET_/, /^AF_/, /^SOCK_/, /^IPPROTO_/, /^TCP_/, /^UDP_/, /^IP_/, /^WSAE/
    categories['network'] << const
  when /^CERT_/, /^CRYPT/, /^ALG_/, /^CALG_/, /^PROV_/
    categories['cryptography'] << const
  when /^LANG_/, /^SUBLANG_/, /^SORT_/, /^LOCALE_/
    categories['locale'] << const
  when /^SQL_/, /^ODBC_/
    categories['database'] << const
  when /^PRINTER_/, /^PRINT_/, /^DM_/
    categories['printing'] << const
  when /^MCI_/, /^WAVE_/, /^MIDI_/
    categories['multimedia'] << const
  when /^IMAGE_/, /^PE_/, /^SECTION_/
    categories['pe_format'] << const
  when /^TAPE_/, /^IOCTL_/
    categories['device_io'] << const
  when /^RPC_/, /^NDR_/
    categories['rpc'] << const
  when /^SNMP_/
    categories['snmp'] << const
  when /^HTTP_/, /^INTERNET_/, /^WINHTTP_/
    categories['http_internet'] << const
  when /^SECURITY_/, /^DOMAIN_/, /^POLICY_/
    categories['security_policy'] << const
  else
    categories['misc'] << const
  end
end

# Print categorization results
puts "\nCategorization Results:"
puts "=" * 50

categories.keys.sort.each do |category|
  puts "#{category.upcase}: #{categories[category].length} constants"
end

puts "\nTop 10 categories by size:"
categories.sort_by { |k, v| -v.length }.first(10).each do |category, consts|
  puts "  #{category}: #{consts.length}"
end

# Show some examples from each category
puts "\nExamples from each category:"
categories.keys.sort.each do |category|
  puts "\n#{category.upcase} (#{categories[category].length} total):"
  categories[category].first(5).each do |const|
    puts "  #{const}"
  end
  puts "  ..." if categories[category].length > 5
end